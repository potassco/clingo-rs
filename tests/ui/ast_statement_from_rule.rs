// ast_statement_from_rule
use clingo::*;

fn main() {
    let loc = ast::Location::default();
    let sym = Symbol::create_id("test", true).unwrap();
    let term = ast::symbolic_term(&loc, &sym).unwrap();
    let atom = ast::symbolic_atom(term).unwrap();
    let lit = ast::basic_literal_from_symbolic_atom(&loc, ast::Sign::NoSign, atom).unwrap();
    let hlit: ast::Head = lit.clone().into();
    let blit: ast::BodyLiteral = lit.into();
    let body = vec![blit];
    let rule = ast::rule(&loc, hlit, &body).unwrap();
    let stmt: ast::Statement = rule.into();

    drop(rule);
    let _end = stmt;
}
