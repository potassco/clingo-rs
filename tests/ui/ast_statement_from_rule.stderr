error[E0382]: use of moved value: `rule`
  --> tests/ui/ast_statement_from_rule.rs:16:10
   |
13 |     let rule = ast::rule(&loc, hlit, &body).unwrap();
   |         ---- move occurs because `rule` has type `clingo::ast::Rule<'_>`, which does not implement the `Copy` trait
14 |     let stmt: ast::Statement = rule.into();
   |                                     ------ `rule` moved due to this method call
15 |
16 |     drop(rule);
   |          ^^^^ value used here after move
   |
note: `into` takes ownership of the receiver `self`, which moves `rule`
  --> $RUST/core/src/convert/mod.rs
   |
   |     fn into(self) -> T;
   |             ^^^^
help: you can `clone` the value and consume it, but this might not be your desired behavior
   |
14 |     let stmt: ast::Statement = rule.clone().into();
   |                                     ++++++++
